trieste/utils/misc.py

```diff
@@ -55,3 +57,114 @@ def to_numpy(t: TensorType) -> np.ndarray:
         return t.numpy()
 
     return t
+
+
+T_co = TypeVar("T_co", covariant=True)
+""" An unbounded covariant type variable. """
+
+
+class Result(Generic[T_co], ABC):
+    """
+    Represents the result of an operation that can fail with an exception. It contains either the
+    operation return value (in an :class:`Ok`), or the exception raised (in an :class:`Err`).
+
+    To check whether instances such as
+
+        >>> res = Ok(1)
+        >>> other_res = Err(ValueError("whoops"))
+
+    contain a value, use :attr:`is_ok` (or :attr:`is_err`)
+
+        >>> res.is_ok
+        True
+        >>> other_res.is_ok
+        False
+
+    We can access the value if it :attr:`is_ok` using :meth:`unwrap`.
+
+        >>> res.unwrap()
+        1
+
+    Trying to access the value of a failed :class:`Result`, or :class:`Err`, will raise the wrapped
+    exception
+
+        >>> other_res.unwrap()
+        Traceback (most recent call last):
+            ...
+        ValueError: whoops
+
+    **Note:** This class is not intended to be subclassed other than by :class:`Ok` and
+    :class:`Err`.
+    """
+
+    @property
+    @abstractmethod
+    def is_ok(self) -> bool:
+        """ `True` if this :class:`Result` contains a value, else `False`. """
+
+    @property
+    def is_err(self) -> bool:
+        """
+        `True` if this :class:`Result` contains an error, else `False`. The opposite of
+        :py:attr:`is_ok`.
+        """
+        return not self.is_ok
+
+    @abstractmethod
+    def unwrap(self) -> T_co:
+        """
+        :return: The contained value, if it exists.
+        :raise Exception: If there is no contained value.
+        """
+
+
+@final
+class Ok(Result[T_co]):
+    """ Wraps the result of a successful evaluation. """
+
+    def __init__(self, value: T_co):
+        """
+        :param value: The result of a successful evaluation.
+        """
+        self._value = value
+
+    def __repr__(self) -> str:
+        """"""
```

Comment:
```suggestion
```
?
(Or
```suggestion
        ""
```
at least looks more like 'empty string' to me)

Reply:
i added this because the docs inherit `object.__repr__`'s docstring, which doesn't anything particularly useful and appears all over the docs. This overrides that (so that `__repr__` doesn't appear in the docs). As for `""""""`, PEP-257 says docstrings should always be triple-double quotes, and I expect some linters will flag up anything else

Reply:
Ah, I didn't know the docs finally handle inheriting docstrings (I remember having found the lack of this quite frustrating in the past-- I even wrote a metaclass that copies docstrings from the superclass if not overwritten...). In that case, it might be helpful for `is_ok` to NOT overwrite the docstring, and get the generic one from the parent class?

Reply:
this is using sphinx-autoapi which is a different engine to what you're using

Reply:
i'd prefer to be specific in the subclasses since we can

Reply:
note the users are more likely to read the docs for `Result` than its implementations